File: content/tracks/dsa-typescript/concepts/10-tree-bfs/tree-bfs_test.ts
import { TreeNode, levelOrder, rightSideView, zigzagLevelOrder } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) throw new Error(`expected ${b}, got ${a}`);
}

function buildTree(vals: (number | null)[]): TreeNode | null {
  if (vals.length === 0 || vals[0] === null) return null;
  const root = new TreeNode(vals[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length > 0 && i < vals.length) {
    const node = queue.shift()!;
    if (i < vals.length && vals[i] !== null) {
      node.left = new TreeNode(vals[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < vals.length && vals[i] !== null) {
      node.right = new TreeNode(vals[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

// --- levelOrder ---

test("levelOrder: 3-level tree", () => {
  assertEqual(levelOrder(buildTree([3, 9, 20, null, null, 15, 7])), [[3], [9, 20], [15, 7]]);
});

test("levelOrder: single node", () => {
  assertEqual(levelOrder(new TreeNode(1)), [[1]]);
});

test("levelOrder: null", () => {
  assertEqual(levelOrder(null), []);
});

test("levelOrder: two levels", () => {
  assertEqual(levelOrder(buildTree([1, 2, 3])), [[1], [2, 3]]);
});

test("levelOrder: left-skewed", () => {
  const t = new TreeNode(1, new TreeNode(2, new TreeNode(3)));
  assertEqual(levelOrder(t), [[1], [2], [3]]);
});

// --- rightSideView ---

test("rightSideView: standard case", () => {
  assertEqual(rightSideView(buildTree([1, 2, 3, null, 5, null, 4])), [1, 3, 4]);
});

test("rightSideView: single node", () => {
  assertEqual(rightSideView(new TreeNode(1)), [1]);
});

test("rightSideView: null", () => {
  assertEqual(rightSideView(null), []);
});

test("rightSideView: left-only subtree visible at deepest level", () => {
  const t = new TreeNode(1, new TreeNode(2, new TreeNode(3)));
  assertEqual(rightSideView(t), [1, 2, 3]);
});

test("rightSideView: right child extends deeper", () => {
  assertEqual(rightSideView(buildTree([1, 2, 3, null, 5])), [1, 3, 5]);
});

// --- zigzagLevelOrder ---

test("zigzagLevelOrder: 3-level tree", () => {
  assertEqual(
    zigzagLevelOrder(buildTree([3, 9, 20, null, null, 15, 7])),
    [[3], [20, 9], [15, 7]]
  );
});

test("zigzagLevelOrder: single node", () => {
  assertEqual(zigzagLevelOrder(new TreeNode(1)), [[1]]);
});

test("zigzagLevelOrder: null", () => {
  assertEqual(zigzagLevelOrder(null), []);
});

test("zigzagLevelOrder: two levels", () => {
  // Level 0 (L→R): [1]; Level 1 (R→L): [3, 2]
  assertEqual(zigzagLevelOrder(buildTree([1, 2, 3])), [[1], [3, 2]]);
});

test("zigzagLevelOrder: three levels full tree", () => {
  // Level 0 (L→R): [1]; Level 1 (R→L): [3,2]; Level 2 (L→R): [4,5,6,7]
  const t = buildTree([1, 2, 3, 4, 5, 6, 7]);
  assertEqual(zigzagLevelOrder(t), [[1], [3, 2], [4, 5, 6, 7]]);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/dsa-python/concepts/10-tree-bfs/tree-bfs_test.py
import sys
from stub import TreeNode, level_order, right_side_view, zigzag_level_order, _build

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# level_order
r = level_order(_build([3, 9, 20, None, None, 15, 7]))
test("level_order: standard", r == [[3], [9, 20], [15, 7]], f"got {r}")
test("level_order: empty tree", level_order(None) == [])
r = level_order(_build([1]))
test("level_order: single node", r == [[1]], f"got {r}")
r = level_order(_build([1, 2, 3, 4, 5]))
test("level_order: three levels", r == [[1], [2, 3], [4, 5]], f"got {r}")
n1 = TreeNode(1); n2 = TreeNode(2); n3 = TreeNode(3)
n1.right = n2; n2.right = n3
r = level_order(n1)
test("level_order: right-skewed", r == [[1], [2], [3]], f"got {r}")

# right_side_view
r = right_side_view(_build([1, 2, 3, None, 5, None, 4]))
test("right_view: standard", r == [1, 3, 4], f"got {r}")
r = right_side_view(_build([1, None, 3]))
test("right_view: right only", r == [1, 3], f"got {r}")
test("right_view: empty", right_side_view(None) == [])
r = right_side_view(_build([1]))
test("right_view: single", r == [1], f"got {r}")
r = right_side_view(_build([1, 2, 3, 4]))
test("right_view: deeper left visible", r == [1, 3, 4], f"got {r}")

# zigzag_level_order
r = zigzag_level_order(_build([3, 9, 20, None, None, 15, 7]))
test("zigzag: standard", r == [[3], [20, 9], [15, 7]], f"got {r}")
r = zigzag_level_order(_build([1]))
test("zigzag: single node", r == [[1]], f"got {r}")
test("zigzag: empty", zigzag_level_order(None) == [])
r = zigzag_level_order(_build([1, 2, 3, 4, 5, 6, 7]))
test("zigzag: 3 levels", r == [[1], [3, 2], [4, 5, 6, 7]], f"got {r}")
r = zigzag_level_order(_build([1, 2, 3]))
test("zigzag: 2 levels", r == [[1], [3, 2]], f"got {r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)
